### PR TITLE
token-swap: Support two token programs at once

### PR DESCRIPTION
**Pull request by joncinque**
#### Problem

Token-2022 is out on all networks, but it's not clear how to integrate it in on-chain programs.

#### Solution

Add some docs with a concrete example using token-swap.  This only covers the simplest migration towards supporting both token and token-2022, but not both in the same instruction. That'll happen in a separate PR.

@jacobcreech no need to look at the program changes, but any comments on the doc are extremely appreciated :smile: 

This PR will be merged without squash so that we can link all of the commits in the doc aftewards.

### CODE REVIEW

File: docs/src/token-2022/onchain.md [hunk 1]
```diff
@@ -2,4 +2,229 @@
...
+$ solana program dump TokenzQdBNbLqP5VEhdkAS6EPFLC1PHnBqCXEpPxuEb spl_token_2022.so
+```
+
+If you're using `solana-test-validator` for your tests, you can include it using:
+
+```console
+$ solana-test-validator -c TokenzQdBNbLqP5VEhdkAS6EPFLC1PHnBqCXEpPxuEb 
+```
+
+Note: This step is temporary, until Token-2022 is included by default in
```
Comment (jacobcreech) [inline docs/src/token-2022/onchain.md:86]:
```suggestion
**Note**: This step is temporary, until Token-2022 is included by default in
```

**Review by jacobcreech — COMMENTED** [review]
Good guide! I added a few comments, feel free to ignore them though. Looking forward to seeing more examples of using token-2022

File: docs/src/token-2022/onchain.md [hunk 2]
```diff
@@ -2,4 +2,229 @@
 title: On-chain Program Guide
 ---
 
-Coming soon!
+This guide is meant for on-chain program / dapp developers who want to support
+Token and Token-2022 concurrently.
+
+## Prerequisites
+
+This guide requires the Solana CLI tool suite.
```
[the remark this answers is not in the document]
Comment (joncinque) [reply docs/src/token-2022/onchain.md:10]: Great call, yes it's 1.10.33 for now

File: docs/src/token-2022/onchain.md [hunk 3]
```diff
@@ -2,4 +2,229 @@
...
+## Motivation
+
+On-chain program developers are accustomed to only including one token program,
+to be used for all tokens in the application.
+
+With the addition of Token-2022, developers must update on-chain programs. This
+guide walks through the steps required to support both.
+
+Important note: if you do not wish to support Token-2022, there is nothing to do.
+Your existing on-chain program will loudly fail if an instruction includes any
```
[the remark this answers is not in the document]
Comment (joncinque) [reply docs/src/token-2022/onchain.md:21]: Yeah most likely. I'll add that in

**Comment by joncinque** [comment]
Ok, this should be good for another pass

**Review by jacobcreech — APPROVED** [review]
Looks good on my end!

**Pull request merged**